Mathlib/RingTheory/Ideal/IsPrincipal.lean

```diff
@@ -0,0 +1,299 @@
+  rw [hf, one_pow] at this
+  refine' span_le.mpr _ this
+  rintro _ hx
+  simp_rw [Set.mem_image] at hx
+  rcases hx with ⟨x, _, rfl⟩
+  have : span ({(x:R) ^ (n + 1)} : Set R) ≤ span ((fun x : R => x ^ (n + 1)) '' s) := by
+    rw [span_le, Set.singleton_subset_iff]
+    exact subset_span ⟨x, x.prop, rfl⟩
+  refine this ?_
+  rw [mul_pow, mem_span_singleton]
+  exact ⟨f x ^ (n + 1), mul_comm _ _⟩
+#align ideal.span_pow_eq_top Ideal.span_pow_eq_top
+
+theorem span_singleton_mul_span_singleton (r s : R) :
+    span {r} * span {s} = (span {r * s} : Ideal R) := by
+  unfold span
+  rw [Submodule.span_mul_span, Set.singleton_mul_singleton]
+#align ideal.span_singleton_mul_span_singleton Ideal.span_singleton_mul_span_singleton
+
+theorem span_singleton_pow (s : R) (n : ℕ) : span {s} ^ n = (span {s ^ n} : Ideal R) := by
+  induction' n with n ih; · simp [Set.singleton_one]
+  simp only [pow_succ, ih, span_singleton_mul_span_singleton]
+#align ideal.span_singleton_pow Ideal.span_singleton_pow
+
+end CommSemiring
+
+section Ring
+
+variable [Ring R]
+
+@[simp]
+theorem span_singleton_neg (x : R) : (span {-x} : Ideal R) = span {x} := by
+  ext
+  simp only [mem_span_singleton']
+  exact ⟨fun ⟨y, h⟩ => ⟨-y, h ▸ neg_mul_comm y x⟩, fun ⟨y, h⟩ => ⟨-y, h ▸ neg_mul_neg y x⟩⟩
+#align ideal.span_singleton_neg Ideal.span_singleton_neg
+
+end Ring
+
+section IsDomain
+
+variable [CommRing R] [IsDomain R]
+
+variable (R) in
+/-- The equivalence between `Associates R` and the principal ideals of `R` defined by sending the
+class of `x` to the principal ideal generated by `x`. -/
+noncomputable def associatesEquivIsPrincipal :
+    Associates R ≃ {I : Ideal R // Submodule.IsPrincipal I} := by
+  refine Equiv.ofBijective (fun x ↦ ⟨Ideal.span {Quot.out x}, ⟨Quot.out x, rfl⟩⟩)
+    ⟨fun  _ _ h ↦ ?_, fun ⟨I, hI⟩ ↦ ?_⟩
+  · rw [Subtype.mk_eq_mk, span_singleton_eq_span_singleton] at h
+    exact Quotient.out_equiv_out.mp h
+  · obtain ⟨x, hx⟩ := hI
+    refine ⟨⟦x⟧, ?_⟩
+    rw [Subtype.mk_eq_mk, hx, submodule_span_eq, span_singleton_eq_span_singleton]
+    exact Associates.mk_quot_out x
+
+@[simp]
+theorem associatesEquivIsPrincipal_apply (x : R) :
+    associatesEquivIsPrincipal R ⟦x⟧ = Ideal.span {x} := by
+  rw [associatesEquivIsPrincipal, Equiv.ofBijective_apply, span_singleton_eq_span_singleton]
+  exact Associates.mk_quot_out x
+
+end IsDomain
+
+end Ideal
+
+section misc
+
+open Ring Set
+
+variable {R : Type*}
+
+theorem Ideal.factors_decreasing [CommSemiring R] [IsDomain R] (b₁ b₂ : R) (h₁ : b₁ ≠ 0)
+    (h₂ : ¬IsUnit b₂) :
+    span ({b₁ * b₂} : Set R) < span {b₁} :=
+  lt_of_le_not_le
+    (Ideal.span_le.2 <| singleton_subset_iff.2 <| Ideal.mem_span_singleton.2 ⟨b₂, rfl⟩) fun h =>
+    h₂ <| isUnit_of_dvd_one <|
+        (mul_dvd_mul_iff_left h₁).1 <| by rwa [mul_one, ← Ideal.span_singleton_le_span_singleton]
+#align ideal.factors_decreasing Ideal.factors_decreasing
+
+lemma Ideal.isPrime_of_maximally_disjoint  [CommSemiring R] (I : Ideal R) (S : Submonoid R)
```

Comment:
As I mentioned in the description, a certain number of results could not stay in `Basic` since they depend on results that have been moved to `IsPrincipal`. For the moment, I moved those results to `IsPrincipal` as a temporary solution (in the `misc` section). I could put these results in `Operations` but the file is already quite big so I am not sure it's the best way to proceed. I welcome any suggestions on what to do with those...